src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1064,304 +907,250 @@ internal TransitionRegex<S> CreateDerivative()
+                    {
+                        Debug.Assert(_left._left is not null && _left._right is not null);
 
-                case SymbolicRegexNodeKind.Loop:
-                    // d(R*) = d(R+) = d(R)R*
-                    Debug.Assert(_left is not null);
-                    Debug.Assert(_upper > 0);
+                        // This pattern is for the path where the backtracking matcher would find the match from the first alternative
+                        SymbolicRegexNode<S> leftLeftPath = _builder.CreateConcat(_left._left, _right);
 
-                    if (eager && IsLazy && _lower == 0)
-                    {
-                        // This is nothing because the backtracking matcher would prefer to exit the loop
-                        transition = TransitionRegex<S>.Leaf(_builder._nothing);
+                        // The eager derivative of that path will be used when the pattern is nullable, i.e., the backtracking matcher
+                        // would end the match rather than go onto paths in the second alternative. When the path through the first
+                        // alternative is not nullable or the derivative is not eager, then all paths through it are considered.
+                        bool leftLeftEager = eager && leftLeftPath.IsNullableFor(context);
+                        if (leftLeftPath.AddTransitions(elem, context, transitions, continuation, effects, leftLeftEager))
+                            return true;
+                        if (!leftLeftEager)
+                            return _builder.CreateConcat(_left._right, _right).AddTransitions(elem, context, transitions, continuation, effects, eager);
+                        break;
                     }
                     else
                     {
-                        TransitionRegex<S> step = _left.CreateDerivativeWithEffects(eager);
-                        if (IsStar || IsPlus)
+                        // Helper function for the case where the right side consumes the character
+                        static bool RightTransition(SymbolicRegexNode<S> node, S elem, uint context, List<(SymbolicRegexNode<S>, DerivativeEffect[]?)> transitions,
+                            List<SymbolicRegexNode<S>> continuation, Stack<DerivativeEffect>? effects, bool eager)
                         {
-                            transition = step.Concat(_builder.CreateLoop(_left, IsLazy));
+                            Debug.Assert(node._left is not null && node._right is not null);
+                            // Remember current number of effects so that we know how many to pop
+                            int j = effects?.Count ?? 0;
+                            if (effects is not null)
+                            {
+                                // Push all effects onto the effects stack
+                                node._left.ApplyEffects(effect => effects.Push(effect), context);
```

Comment:
The other place where `ApplyEffects` gets used is at the end of `FindEndPositionCapturing`, and generally we may have to call it once per match. In that use a second position parameter also needs to be passed in to the actual effect applying function. I'm not sure what's the best way to optimize for both of these cases without duplicating the code. This derivative construction is pretty allocatey, so probably not a huge problem here, but the other use should be optimized for at least.

At this call site I can actually change the call to just pass in `effects.Push`, since that's already castable to `Action<DerivativeEffect>`, although I imagine that must still incur some allocation to capture the "this" at least.

Reply:
> I'm not sure what's the best way to optimize for both of these cases without duplicating the code

We'd typically pass through a T generic state argument, and that could be a value tuple is there was more than one value to pass through.

Reply:
I went ahead and made that change